In [196]:
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objects as go

In [119]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [120]:
%sql postgresql://postgres:postgres@localhost:5430/eskom_db

In [121]:
%%sql 
DROP TABLE IF EXISTS eskom_data;
CREATE TABLE  eskom_data(
    date_time_hour_beginning DATE,
    original_res_forecast_before_lockdown NUMERIC,
    residual_forecast  NUMERIC,	
    rsa_contracted_forecast	 NUMERIC,
    dispatchable_generation	NUMERIC ,
    residual_demand NUMERIC,
    rsa_contracted_demand  NUMERIC,
    international_exports  NUMERIC,
    international_imports  NUMERIC,
    thermal_generation  NUMERIC,
    nuclear_generation  NUMERIC,
    eskom_gas_generation  NUMERIC,	
    eskom_ocgt_generation  NUMERIC,	
    hydro_water_generation  NUMERIC,
    pumped_water_generation	 NUMERIC,
    ils_usage  NUMERIC,
    manual_load_reduction NUMERIC,
    ios_excl_ils_and_mlr  NUMERIC,
    dispatchable_ipp_ocgt  NUMERIC,
    eskom_gas_sco  NUMERIC,
    eskom_ocgt_sco  NUMERIC,
    hydro_water_sco	 NUMERIC,
    pumped_water_sco  NUMERIC,
    pumping	 NUMERIC,
    wind  NUMERIC,
    pv  NUMERIC,
    csp  NUMERIC,
    other_re  NUMERIC,	
    total_re  NUMERIC,	
    wind_installed_capacity	 NUMERIC,
    pv_installed_capacity  NUMERIC,	
    csp_installed_capacity  NUMERIC,
    other_re_installed_capacity	 NUMERIC,
    total_re_installed_capacity	 NUMERIC,
    installed_eskom_capacity  NUMERIC,
    total_pclf  NUMERIC,
    total_uclf  NUMERIC,
    total_oclf  NUMERIC,
    total_uclf_and_oclf	 NUMERIC,
    non_comm_sentout  NUMERIC,	
    drakensberg_gen_unit_hours  NUMERIC,
    palmiet_gen_unit_hours  NUMERIC,	
    ingula_gen_unit_hours NUMERIC
    
);



 * postgresql://postgres:***@localhost:5430/eskom_db
Done.
Done.


[]

In [122]:
%%sql 
COPY eskom_data(date_time_hour_beginning,original_res_forecast_before_lockdown,residual_forecast,rsa_contracted_forecast,dispatchable_generation,residual_demand,rsa_contracted_demand,international_exports,international_imports,thermal_generation,nuclear_generation,eskom_gas_generation,eskom_ocgt_generation,hydro_water_generation,pumped_water_generation,ils_usage,manual_load_reduction,ios_excl_ils_and_mlr,dispatchable_ipp_ocgt,eskom_gas_sco,eskom_ocgt_sco,hydro_water_sco,pumped_water_sco,pumping,wind,pv,csp,other_re,total_re,wind_installed_capacity,pv_installed_capacity,csp_installed_capacity,other_re_installed_capacity,total_re_installed_capacity,installed_eskom_capacity,total_pclf,total_uclf,total_oclf,total_uclf_and_oclf,non_comm_sentout,drakensberg_gen_unit_hours,palmiet_gen_unit_hours,ingula_gen_unit_hours)
FROM '/Users/ds_learner16/Documents/Eskom/Eskom_Data_Analysis/ESK2033.csv'
DELIMITER ','
CSV HEADER
;

 * postgresql://postgres:***@localhost:5430/eskom_db
37704 rows affected.


[]

In [123]:
%%sql 
DROP TABLE IF EXISTS eskom_clean;
CREATE TABLE eskom_clean 
AS 
SELECT
    EXTRACT(YEAR FROM date_time_hour_beginning) as year,
    EXTRACT(MONTH FROM date_time_hour_beginning) as month,
    EXTRACT(DAY FROM date_time_hour_beginning) as day,
    date_time_hour_beginning,
    original_res_forecast_before_lockdown,
    residual_forecast,
    rsa_contracted_forecast,
    dispatchable_generation,
    residual_demand,
    rsa_contracted_demand,
    international_exports,
    international_imports,
    thermal_generation,
    nuclear_generation,
    eskom_gas_generation,
    eskom_ocgt_generation,
    hydro_water_generation,
    pumped_water_generation,
    ils_usage,
    manual_load_reduction,
    ios_excl_ils_and_mlr,
    dispatchable_ipp_ocgt,
    eskom_gas_sco,
    eskom_ocgt_sco,
    hydro_water_sco,
    pumped_water_sco,
    pumping,
    wind,
    pv,
    csp,
    other_re,
    total_re,
    wind_installed_capacity,
    pv_installed_capacity,
    csp_installed_capacity,
    other_re_installed_capacity,
    total_re_installed_capacity,
    installed_eskom_capacity,
    total_pclf,
    total_uclf,
    total_oclf,
    total_uclf_and_oclf,
    non_comm_sentout,
    drakensberg_gen_unit_hours,
    palmiet_gen_unit_hours,
    ingula_gen_unit_hours

FROM eskom_data;
COPY eskom_clean TO '/Users/ds_learner16/Documents/Eskom/Eskom_Data_Analysis/ESK2033_clean.csv' WITH DELIMITER ',' CSV HEADER;



 * postgresql://postgres:***@localhost:5430/eskom_db
Done.
37704 rows affected.
37704 rows affected.


[]

# Number of Rows and Columns 
count rows = 42824 estimated before clean after 37704
removed forecast dates added average for original res before lockdown

count cols = 41

Look at our variables

In [124]:
all = %sql SELECT * FROM eskom_clean LIMIT 5 
all

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


[(Decimal('2018'), Decimal('4'), Decimal('1'), datetime.date(2018, 4, 1), Decimal('27517.5'), Decimal('19904.967'), Decimal('20367.066'), Decimal('20237'), Decimal('20237'), Decimal('20722.058'), Decimal('1215.902'), Decimal('1120'), Decimal('19444'), Decimal('931'), Decimal('0'), Decimal('0'), Decimal('360'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('-2'), Decimal('-2'), Decimal('0'), Decimal('-1614'), Decimal('454.964'), Decimal('0'), Decimal('17.555'), Decimal('12.539'), Decimal('485.058'), Decimal('2077.72'), Decimal('1474.19'), Decimal('300'), Decimal('21.78'), Decimal('3873.69'), Decimal('44546'), Decimal('3987.472'), Decimal('8028.71'), Decimal('275.907'), Decimal('8304'), Decimal('617'), Decimal('383'), Decimal('81.8'), Decimal('36.9'), Decimal('30.41')),
 (Decimal('2018'), Decimal('4'), Decimal('1'), datetime.date(2018, 4, 1), Decimal('27517.5'), Decimal('19553.899'), Decimal('19988.733'), Decimal('19744'), Decimal('19744'), Decimal('20188.493'), Decimal('1203.474'), Decimal('1106'), Decimal('19297'), Decimal('930'), Decimal('0'), Decimal('0'), Decimal('360'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('-2'), Decimal('-2'), Decimal('0'), Decimal('-1945'), Decimal('392.958'), Decimal('0'), Decimal('39.131'), Decimal('12.404'), Decimal('444.493'), Decimal('2077.72'), Decimal('1474.19'), Decimal('300'), Decimal('21.78'), Decimal('3873.69'), Decimal('44546'), Decimal('3987.472'), Decimal('7727.302'), Decimal('244.907'), Decimal('7972'), Decimal('209'), Decimal('388'), Decimal('83'), Decimal('38.5'), Decimal('32.85')),
 (Decimal('2018'), Decimal('4'), Decimal('1'), datetime.date(2018, 4, 1), Decimal('27517.5'), Decimal('19314.284'), Decimal('19731.239'), Decimal('19631'), Decimal('19631'), Decimal('20019.603'), Decimal('1177.571'), Decimal('1117'), Decimal('19165'), Decimal('931'), Decimal('0'), Decimal('0'), Decimal('360'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('-2'), Decimal('-2'), Decimal('0'), Decimal('-1938'), Decimal('337.358'), Decimal('0'), Decimal('39.163'), Decimal('12.082'), Decimal('388.603'), Decimal('2077.72'), Decimal('1474.19'), Decimal('300'), Decimal('21.78'), Decimal('3873.69'), Decimal('44546'), Decimal('3987.472'), Decimal('7704.704'), Decimal('193.727'), Decimal('7898'), Decimal('431'), Decimal('388'), Decimal('83.8'), Decimal('40.3'), Decimal('35.6')),
 (Decimal('2018'), Decimal('4'), Decimal('1'), datetime.date(2018, 4, 1), Decimal('27517.5'), Decimal('19342.679'), Decimal('19753.554'), Decimal('19731'), Decimal('19731'), Decimal('20079.454'), Decimal('1184.312'), Decimal('1118'), Decimal('19279'), Decimal('930'), Decimal('0'), Decimal('0'), Decimal('360'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('-2'), Decimal('-2'), Decimal('0'), Decimal('-1952'), Decimal('297.234'), Decimal('0'), Decimal('38.942'), Decimal('12.278'), Decimal('348.454'), Decimal('2077.72'), Decimal('1474.19'), Decimal('300'), Decimal('21.78'), Decimal('3873.69'), Decimal('44546'), Decimal('3990.072'), Decimal('7702.868'), Decimal('187'), Decimal('7889'), Decimal('868'), Decimal('389'), Decimal('85'), Decimal('42'), Decimal('37.76')),
 (Decimal('2018'), Decimal('4'), Decimal('1'), datetime.date(2018, 4, 1), Decimal('27517.5'), Decimal('19538.89'), Decimal('19988.365'), Decimal('19890'), Decimal('19890'), Decimal('20237.49'), Decimal('1197.271'), Decimal('1108'), Decimal('19369'), Decimal('930'), Decimal('0'), Decimal('0'), Decimal('360'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('-2'), Decimal('-2'), Decimal('0'), Decimal('-1873'), Decimal('309.978'), Decimal('0'), Decimal('25.024'), Decimal('12.488'), Decimal('347.49'), Decimal('2077.72'), Decimal('1474.19'), Decimal('300'), Decimal('21.78'), Decimal('3873.69'), Decimal('44546'), Decimal('3990.472'), Decimal('7685.115'), Decimal('187'), Decimal('7872'), Decimal('115'), Decimal('385'), Decimal('85.

## Comparing 

dispatchable_generation

dispatchable_ipp_ocgt




international_exports 

international_imports




thermal 

nuclear

eskom_gas

eskom_ocgt

pumped_water

wind 

pv

csp

other_re




drakensburg

palmiet

ingula

In [125]:
sum_dispatchable = %sql SELECT year as "year",SUM(dispatchable_generation) as "Dispatchable_generation(in GW)",SUM(dispatchable_ipp_ocgt) as  "Dispatchable IPP OCGT(in GW)" FROM eskom_clean GROUP BY year ORDER BY year
sum_dispatchable

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


[(Decimal('2018'), Decimal('169700362.829'), Decimal('377118.581')),
 (Decimal('2019'), Decimal('219562829.479'), Decimal('610770.575')),
 (Decimal('2020'), Decimal('206725336.016'), Decimal('661405.032')),
 (Decimal('2021'), Decimal('210021594.912'), Decimal('976052.052')),
 (Decimal('2022'), Decimal('114527886.147'), Decimal('509943.467'))]

## How much energy is generated yearly by eskom and other enties in the country

In [136]:
fig = px.bar(sum_dispatchable, x="year", y ="Dispatchable_generation(in GW)" ,title="Total of Dispatchable Generation Yearly" )
fig1 = px.bar(sum_dispatchable,x="year" ,y="Dispatchable IPP OCGT(in GW)",title="Total of Dispatchable IPP OCGT Generation Yearly")
fig.show()
fig1.show()

## Do we import more energy or export more energy?

In [139]:
imports_vs_exports = %sql SELECT year AS "year",SUM(international_imports) AS "imports",SUM(international_exports)  AS "exports" FROM eskom_clean  GROUP BY year ORDER BY year
imports_vs_exports

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


[(Decimal('2018'), Decimal('7600760'), Decimal('10638946.468')),
 (Decimal('2019'), Decimal('9829822.4'), Decimal('14995379.056')),
 (Decimal('2020'), Decimal('9888997'), Decimal('13653406.142')),
 (Decimal('2021'), Decimal('10151824'), Decimal('13708931.021')),
 (Decimal('2022'), Decimal('6074617'), Decimal('6805214.404'))]

In [198]:
fig4 = px.line(imports_vs_exports,x = "year",y=["imports","exports"],title="Total Imports VS Exports Yearly(in GW)")
fig4.show()


In [141]:
fig5 = px.box(imports_vs_exports,y=["imports","exports"],title="Boxplot Imports and Exports")
fig5.show()

In [161]:
energy_sources = %sql SELECT year as "year",AVG(thermal_generation) as "thermal",AVG(nuclear_generation) as "nuclear",AVG(eskom_gas_generation) as "eskom_gas",AVG(eskom_ocgt_generation) as "eskom_ocgt",AVG(pumped_water_generation) as "pumped water",AVG(wind) as "wind",AVG(pv) as "pv",AVG(csp) as "csp",AVG(other_re) as "other_re" FROM eskom_clean  GROUP BY year ORDER BY year
energy_sources

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


[(Decimal('2018'), Decimal('23335.488006363636'), Decimal('1151.2146969696969697'), Decimal('0.36878787878787878788'), Decimal('64.6908206060606061'), Decimal('544.9339393939393939'), Decimal('364.5652481818181818'), Decimal('115.4267080303030303'), Decimal('12.8029777272727273'), Decimal('1231.4713956060606061')),
 (Decimal('2019'), Decimal('22250.353296118721'), Decimal('1550.8905251141552511'), Decimal('0.18641552511415525114'), Decimal('175.5685045662100457'), Decimal('563.8417808219178082'), Decimal('379.5649978310502283'), Decimal('177.7569319634703196'), Decimal('9.1509718036529680'), Decimal('1322.7106149543378995')),
 (Decimal('2020'), Decimal('20993.572397427140'), Decimal('1310.8622495446265938'), Decimal('0.36885245901639344262'), Decimal('142.4160109289617486'), Decimal('583.3568989071038251'), Decimal('471.3356622267759563'), Decimal('185.1149203096539162'), Decimal('9.8602550091074681'), Decimal('1420.6174861111111111')),
 (Decimal('2021'), Decimal('21082.486970319635'), Decimal('1388.2902968036529680'), Decimal('0.03242009132420091324'), Decimal('255.1918949771689498'), Decimal('559.5926940639269406'), Decimal('578.6696744292237443'), Decimal('189.0429831050228311'), Decimal('14.1485195205479452'), Decimal('1736.1104408675799087')),
 (Decimal('2022'), Decimal('20964.593354791667'), Decimal('968.4904166666666667'), Decimal('0.55354166666666666667'), Decimal('275.7426000000000000'), Decimal('524.1600000000000000'), Decimal('526.5613014583333333'), Decimal('147.2490760416666667'), Decimal('20.8730152083333333'), Decimal('1700.3761529166666667'))]

## Which one of the energy generation methods yeild the most power in GW

In [186]:
fig6 = px.line(energy_sources,x="year",y=["eskom_gas","eskom_ocgt","pumped water","nuclear","wind","pv","csp","other_re"],title="Avg of energy generation from different sources")
fig6.show()

### Thermal power produces the most power but is showing decline in recent years

In [187]:
fig7 = px.line(energy_sources,x="year",y="thermal",title="Thermal energy generation over the years")
fig7.show()

In [193]:
power_stations = %sql SELECT year,SUM(drakensberg_gen_unit_hours) as "Drakensburg",SUM(palmiet_gen_unit_hours) as "Palmiet",SUM(ingula_gen_unit_hours) as "Ingula" FROM eskom_clean GROUP BY year ORDER BY year
power_stations

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


[(Decimal('2018'), Decimal('540653.025'), Decimal('273389.88'), Decimal('270253.405')),
 (Decimal('2019'), Decimal('696592.25'), Decimal('351267.94'), Decimal('336347.620')),
 (Decimal('2020'), Decimal('752133.90'), Decimal('393694.64'), Decimal('296108.366')),
 (Decimal('2021'), Decimal('730552.25'), Decimal('383171.803'), Decimal('332536.312')),
 (Decimal('2022'), Decimal('398131.982'), Decimal('214183.25'), Decimal('173774.438'))]

## Which is the highest performing power station out of the 3 stations

In [197]:
fig8 = px.line(power_stations,x="year",y=["Drakensburg","Palmiet","Ingula"],title="Power Station Energy generation over the years")
fig8.show()